jumanji/environments/routing/sokoban/env.py

```diff
@@ -0,0 +1,587 @@
+        """
+
+        generator_key, key = jax.random.split(key)
+
+        state = self.generator(generator_key)
+
+        timestep = restart(
+            self._state_to_observation(state),
+            extras=self._get_extras(state),
+        )
+
+        return state, timestep
+
+    def step(
+        self, state: State, action: chex.Array
+    ) -> Tuple[State, TimeStep[Observation]]:
+        """
+        Executes one timestep of the environment's dynamics.
+
+        Args:
+            state: 'State' object representing the current state of the
+            environment.
+            action: Array (int32) of shape ().
+                - 0: move up.
+                - 1: move down.
+                - 2: move left.
+                - 3: move right.
+
+        Returns:
+            state, timestep: next state of the environment and timestep to be
+            observed.
+        """
+
+        # switch to noop if action will have no impact on variable grid
+        action = self.detect_noop_action(
+            state.variable_grid, state.fixed_grid, action, state.agent_location
+        )
+
+        next_variable_grid, next_agent_location = jax.lax.cond(
+            jnp.all(action == -1),
+            lambda: (state.variable_grid, state.agent_location),
+            lambda: self.move_agent(state.variable_grid, action, state.agent_location),
+        )
+
+        next_state = State(
+            key=state.key,
+            fixed_grid=state.fixed_grid,
+            variable_grid=next_variable_grid,
+            agent_location=next_agent_location,
+            step_count=state.step_count + 1,
+        )
+
+        target_reached = self.level_complete(next_state)
+        time_limit_exceeded = next_state.step_count >= self.time_limit
+
+        #done = target_reached | time_limit_exceeded
+        done = jnp.logical_or(target_reached, time_limit_exceeded)
+
+        reward = jnp.asarray(self.reward_fn(state, action, next_state), float)
+
+        observation = self._state_to_observation(next_state)
+
+        extras = self._get_extras(next_state)
+
+        timestep = jax.lax.cond(
+            done,
+            lambda: termination(
+                reward=reward,
+                observation=observation,
+                extras=extras,
+            ),
+            lambda: transition(
+                reward=reward,
+                observation=observation,
+                extras=extras,
+            ),
+        )
+
+        return next_state, timestep
+
+    def observation_spec(self) -> specs.Spec[Observation]:
+        """
+        Returns the specifications of the observation of the `Sokoban`
+        environment.
+
+        Returns:
+            specs.Spec[Observation]: The specifications of the observations.
+        """
+        grid = specs.BoundedArray(
+            shape=(self.num_rows, self.num_cols, 2),
+            dtype=jnp.uint8,
+            minimum=0,
+            maximum=4,
```

Comment:
Just checking, can grid values be 5 or 6?
cf in constants.py:
```
TARGET_AGENT = 5
TARGET_BOX = 6
```

Reply:
this is purely used for the visualisation. Agent on target for example if defined by the presence of the target in the fixed grid and the agent in the variable grid. The unique combination id used used to choose the correct image to display this.